### PR TITLE
LUCENE-10436: Remove deprecated DocValuesFieldExistsQuery, NormsFieldExistsQuery and KnnVectorFieldExistsQuery

### PR DESCRIPTION
**Pull request by zacharymorn**
# Description

Remove deprecated DocValuesFieldExistsQuery, NormsFieldExistsQuery and KnnVectorFieldExistsQuery

# Tests

Passed existing tests.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/lucene/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

### CODE REVIEW

**Comment by zacharymorn** [comment]
For the change entry, I assume this should go into version `10.0.0`?

File: lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java [hunk 1]
```diff
@@ -58,12 +58,6 @@ private static boolean shouldNeverCache(Query query) {
       return true;
     }
 
-    if (query instanceof DocValuesFieldExistsQuery) {
-      // We do not bother caching DocValuesFieldExistsQuery queries since they are already plenty
-      // fast.
-      return true;
-    }
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java:65]: We should have a similar instanceof check for `FieldExistsQuery`?

**Review by jpountz — APPROVED** [review]
Great. We should backport these changes but the actual removals to 9.x to address deprecation warnings.

**Comment by jpountz** [comment]
> For the change entry, I assume this should go into version 10.0.0?

Yes, we need a CHANGES entry under 10.0.0 and a new entry in `lucene/MIGRATE.txt` that recommends replacing `DocValueFieldExistsQuery` and others with `FieldExistsQuery`.

File: lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java (hunk 1 above)
Comment (zacharymorn) [reply lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java:65]: Oh sorry I should have added a nocommit for this. Given `FieldExistsQuery` now supports norms and vectors in addition to doc values, would not caching for also norms and vectors here hurt performance, if we were to have similar instance of check for `FieldExistsQuery`? I'm also wondering if there's a luceneutil like benchmark for these as well?

**Comment by zacharymorn** [comment]
> > For the change entry, I assume this should go into version 10.0.0?
> 
> Yes, we need a CHANGES entry under 10.0.0 and a new entry in `lucene/MIGRATE.txt` that recommends replacing `DocValueFieldExistsQuery` and others with `FieldExistsQuery`.

Sounds good. Added.

**Comment by zacharymorn** [comment]
> Great. We should backport these changes but the actual removals to 9.x to address deprecation warnings.

Thanks for the review! I've created the backporting PR for 9.x here https://github.com/apache/lucene/pull/798

File: lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java (hunk 1 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java:65]: I feel good about not having a benchmark for this. The reasoning is that if the index has a data structure that supports running the query very efficiently, then we should just use it and skip caching. And we have this for doc values, norms and vectors. In contrast, boolean queries for instance need to reconcile multiple queries together, which has overhead.

So +1 to exclude FieldExistsQuery from caching entirely.

Comment (zacharymorn) [reply lucene/core/src/java/org/apache/lucene/search/UsageTrackingQueryCachingPolicy.java:65]: Makes sense. I put it back in https://github.com/apache/lucene/pull/790/commits/1f428d50e78d27f4322e54d7170c66dff64d14a0

**Review by jpountz — APPROVED** [review]

**Comment by zacharymorn** [comment]
Thanks @jpountz !

**Pull request merged**